pypose/optim/kernel.py

```diff
@@ -123,3 +123,149 @@ def forward(self, input: Tensor) -> Tensor:
+        >>> kernel(input)
+        tensor([0.0000, 0.4495, 0.8284, 1.4641, 2.0000])
+    """
+    def __init__(self, delta: float = 1.0) -> None:
+        super().__init__()
+        assert delta > 0, ValueError("delta has to be positive: {}".format(delta))
+        self.delta1 = delta
+        self.delta2 = delta**2
+
+    def forward(self, input: Tensor) -> Tensor:
+        '''
+        Args:
+            input (torch.Tensor): the input tensor (non-negative).
+        '''
+        assert torch.all(input >= 0), 'input has to be non-negative'
+        return 2*(self.delta1*(1/self.delta2+input).sqrt()-1)
+
+
+class Arctan(nn.Module):
+    r"""The robust Arctan kernel cost function.
+
+    .. math::
+        \bm{y}_i=\delta ^{2}\arctan \left ( \frac{\bm{x}_i}{\delta ^{2}}\right )
+
+    where :math:`\delta` (delta) is a hyperparameter, :math:`\bm{x}` and :math:`\bm{y}` are the
+    input and output tensors, respectively.
+
+    Args:
+        delta (float, optional): Specify the Arctan cost. The value must be positive. Default: 1.0
+
+    Note:
+        The input has to be a non-negative tensor and the output tensor has the same shape with the
+        input.
+
+    Example:
+        >>> import pypose.optim.kernel as ppok
+        >>> kernel = ppok.Arctan()
+        >>> input = torch.tensor([0, 0.5, 1, 2, 3])
+        >>> kernel(input)
+        tensor([0.0000, 0.4636, 0.7854, 1.1071, 1.2490])
+    """
+    def __init__(self, delta: float = 1.0) -> None:
+        super().__init__()
+        assert delta > 0, ValueError("delta has to be positive: {}".format(delta))
+        self.delta2 = delta**2
+
+    def forward(self, input: Tensor) -> Tensor:
+        '''
+        Args:
+            input (torch.Tensor): the input tensor (non-negative).
+        '''
+        assert torch.all(input >= 0), 'input has to be non-negative'
+        return self.delta2*(input/self.delta2).arctan()
+
+
+class Tolerant(nn.Module):
+    r"""The robust Tolerant kernel cost function.
+
+    .. math::
+        \bm{y}_i = b\log (1+e^{\frac{\bm{x}_i-a}{b}})-b\log (1+e^{\frac{-a}{b}})
+
+    where :math:`\bm{a}`  and :math:`\bm{b}` are hyperparameters, :math:`\bm{x}` and :math:`\bm{y}` are the
+    input and output tensors, respectively.
+
+    Args:
+        a (float, optional): Specify the Tolerant cost. The value must be positive. Default: 1.0
+        b (float, optional): Specify the Tolerant cost. The value must be positive. Default: 1.0
+
+    Note:
+        The input has to be a non-negative tensor and the output tensor has the same shape with
+        the input.
+
+    Example:
+        >>> import pypose.optim.kernel as ppok
+        >>> kernel = ppok.Tolerant()
+        >>> input = torch.tensor([0, 0.5, 1, 2, 3])
+        >>> kernel(input)
+        tensor([0.0000, 0.4636, 0.7854, 1.1071, 1.2490])
+    """
+    def __init__(self, a: float=1.0, b: float=1.0) -> None:
+        super().__init__()
+        self.a = a
+        self.b = b
```

Comment:
You should verify the values of `a` and `b` to make sure they are in the corrected range.